File: implementations/php/tests/e2e_test.php
<?php
/**
 * LUMEN e2e cross-implementation test — PHP.
 *
 * Run: php tests/e2e_test.php
 */

declare(strict_types=1);

$autoloadPath = __DIR__ . '/../vendor/autoload.php';
if (file_exists($autoloadPath)) {
    require $autoloadPath;
} else {
    require __DIR__ . '/../autoload_manual.php';
}

use Lumen\Compress;
use Lumen\EmptyObject;
use Lumen\Frame;
use Lumen\Hyb128;
use Lumen\ParseResultComplete;

define('VECTORS_PATH', dirname(__DIR__, 3) . '/tests/e2e/shared_vectors.json');
define('GOLDEN_DIR', dirname(__DIR__, 3) . '/tests/e2e/golden');

function loadVectors(): array {
    $raw = file_get_contents(VECTORS_PATH);
    $data = json_decode($raw, false); // no assoc — preserves [] vs {} distinction
    $vectors = [];
    foreach ($data->vectors as $v) {
        $vectors[] = ['name' => $v->name, 'value' => jsonDecodeValue($v->value)];
    }
    return $vectors;
}

/** Convert JSON-decoded value (stdClass for objects) to PHP arrays. */
function jsonDecodeValue(mixed $v): mixed {
    if (is_array($v)) return array_map('jsonDecodeValue', $v);
    if ($v instanceof stdClass) {
        $arr = [];
        foreach ($v as $k => $val) {
            $arr[$k] = jsonDecodeValue($val);
        }
        return count($arr) === 0 ? new EmptyObject() : $arr;
    }
    return $v;
}

function loadGolden(string $name): ?string {
    $path = GOLDEN_DIR . '/' . $name . '.lumen';
    return file_exists($path) ? file_get_contents($path) : null;
}

function loadGoldenFrame(string $name): ?string {
    $path = GOLDEN_DIR . '/' . $name . '.frame';
    return file_exists($path) ? file_get_contents($path) : null;
}

function jsonEqual(mixed $a, mixed $b): bool {
    return json_encode($a, JSON_UNESCAPED_UNICODE) === json_encode($b, JSON_UNESCAPED_UNICODE);
}

/** Encode JSON matching Python's json.dumps default format (spaces after : and ,). */
function jsonDumpsPy(mixed $value): string {
    $json = json_encode($value, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
    // Python's json.dumps uses separators=(', ', ': ') — add space after : and ,
    // Match ':' not followed by whitespace (covers "key":"val", "key":1, "key":{)
    $json = preg_replace('/":(?!\s)/', '": ', $json);
    // Match ',' not followed by whitespace (covers ...",... and ...1,... and ...},...)
    $json = preg_replace('/,(?!\s)/', ', ', $json);
    return $json;
}

function valueEquals(mixed $a, mixed $b): bool {
    if ($a === null && $b === null) return true;
    if ($a instanceof EmptyObject && $b instanceof EmptyObject) return true;
    if (is_float($a) && is_float($b)) return abs($a - $b) < 1e-12;
    // EmptyObject and [] are semantically equivalent (empty collections)
    if ($a instanceof EmptyObject && is_array($b) && $b === []) return true;
    if ($b instanceof EmptyObject && is_array($a) && $a === []) return true;
    if (is_array($a) && is_array($b)) return jsonEqual($a, $b);
    return $a === $b;
}

$SKIP_BINARY_COMPARE = ['float_zero', 'object_empty'];

$passed = 0;
$failed = 0;

function test(string $vector, string $testName, bool $ok, ?string $error = null): void {
    global $passed, $failed;
    if ($ok) { $passed++; } else { $failed++; }
    $status = $ok ? "✓" : "✗";
    $errStr = $error ? ": $error" : "";
    printf("  %s %s — %s%s\n", $status, $testName, $vector, $errStr);
}

// ═══ 1. Compress Roundtrip ═══
echo "\n1. Compress Roundtrip\n";
$vectors = loadVectors();
foreach ($vectors as $v) {
    $name = $v['name'];
    $value = $v['value'];
    try {
        $compressed = Compress::compress($value);
        $decompressed = Compress::decompress($compressed);
        $ok = valueEquals($decompressed, $value);
        test($name, 'compress_roundtrip', $ok);
    } catch (\Throwable $e) {
        test($name, 'compress_roundtrip', false, $e::class . ': ' . $e->getMessage());
    }
}

// ═══ 2. Binary Compatibility ═══
echo "\n2. Binary Compatibility (PHP vs Python golden)\n";
foreach ($vectors as $v) {
    $name = $v['name'];
    $value = $v['value'];
    if (in_array($name, $SKIP_BINARY_COMPARE)) {
        test($name, 'binary_match', true, 'SKIP');
        test($name, 'cross_decode', true, 'SKIP');
        continue;
    }
    try {
        $php = Compress::compress($value);
        $golden = loadGolden($name);
        if ($golden !== null) {
            $ok = $php === $golden;
            test($name, 'binary_match', $ok,
                $ok ? null : sprintf("PHP(%dB):%s != PY(%dB):%s",
                    strlen($php), bin2hex($php), strlen($golden), bin2hex($golden)));
        } else {
            test($name, 'binary_match', false, "golden not found");
        }
    } catch (\Throwable $e) {
        test($name, 'binary_match', false, $e::class . ': ' . $e->getMessage());
    }
    try {
        $golden = loadGolden($name);
        if ($golden !== null) {
            $decoded = Compress::decompress($golden);
            $ok = valueEquals($decoded, $value);
            test($name, 'cross_decode', $ok);
        } else {
            test($name, 'cross_decode', false, "golden not found");
        }
    } catch (\Throwable $e) {
        test($name, 'cross_decode', false, $e::class . ': ' . $e->getMessage());
    }
}

// ═══ 3. Binary Stability ═══
echo "\n3. Binary Stability\n";
$stabilityCases = [
    ['null', null], ['bool', true], ['int', 42], ['float', 3.14],
    ['string', 'hello'], ['array', [1, 2, 3]], ['object', ['key' => 'value']],
    ['mcp_init', ['jsonrpc' => '2.0', 'method' => 'initialize']],
];
foreach ($stabilityCases as [$sn, $sv]) {
    try {
        $c1 = Compress::compress($sv);
        $c2 = Compress::compress($sv);
        test("stability_$sn", 'binary_stability', $c1 === $c2);
    } catch (\Throwable $e) {
        test("stability_$sn", 'binary_stability', false, $e::class . ': ' . $e->getMessage());
    }
}

// ═══ 4. Hyb128 Roundtrip ═══
echo "\n4. Hyb128 Roundtrip\n";
$hyb128Cases = [
    [0, 1], [1, 1], [42, 1], [63, 1],
    [64, 3], [255, 3], [1000, 3], [65535, 3],
    [65536, 5], [100000, 5], [1000000, 5],
];
foreach ($hyb128Cases as [$value, $expectedLen]) {
    $name = "hyb128_$value";
    try {
        $encoded = Hyb128::encodeBytes($value);
        $actualLen = strlen($encoded);
        $expectedEnc = Hyb128::encodedLen($value);
        $decoded = Hyb128::decode($encoded, 0);
        $ok = $actualLen === $expectedLen
           && $expectedEnc === $expectedLen
           && $decoded !== null
           && $decoded['value'] === $value
           && $decoded['headerLen'] === $expectedLen;
        test($name, 'hyb128_roundtrip', $ok);
    } catch (\Throwable $e) {
        test($name, 'hyb128_roundtrip', false, $e::class . ': ' . $e->getMessage());
    }
}

// ═══ 5. Frame Roundtrip ═══
echo "\n5. Frame Roundtrip\n";
$payloads = [
    ['empty', ''],
    ['hello', 'hello'],
    ['json_small', jsonDumpsPy(['method' => 'ping'])],
    ['json_mcp', jsonDumpsPy([
        'jsonrpc' => '2.0', 'id' => 1, 'method' => 'initialize',
        'params' => ['protocolVersion' => '2025-06-18'],
    ])],
];
$frameTypes = [
    ['REQUEST', Frame::REQUEST],
    ['RESPONSE', Frame::RESPONSE],
    ['NOTIFY', Frame::NOTIFY],
];
$flagSets = [
    ['none', 0],
    ['compressed', Frame::FLAG_COMPRESSED],
    ['priority', Frame::FLAG_PRIORITY],
];

foreach ($payloads as [$pname, $payload]) {
    foreach ($frameTypes as [$tname, $ftype]) {
        foreach ($flagSets as [$flname, $flags]) {
            $name = "frame_{$tname}_{$flname}_{$pname}";
            try {
                $frame = Frame::buildFrame($ftype, $flags, $payload);
                $wire = $frame->toBytes();
                $result = Frame::parseFrame($wire, 0);
                if ($result instanceof ParseResultComplete) {
                    $ok = $result->frame->frameType === $ftype
                       && $result->frame->flags === $flags
                       && $result->frame->payload === $payload;
                    test($name, 'frame_roundtrip', $ok);
                } else {
                    test($name, 'frame_roundtrip', false, 'not complete: ' . $result::class);
                }
            } catch (\Throwable $e) {
                test($name, 'frame_roundtrip', false, $e::class . ': ' . $e->getMessage());
            }
        }
    }
}

// ═══ 6. Frame Binary Compatibility ═══
echo "\n6. Frame Binary Compatibility\n";
foreach ($payloads as [$pname, $payload]) {
    foreach ($frameTypes as [$tname, $ftype]) {
        foreach ($flagSets as [$flname, $flags]) {
            $name = "frame_{$tname}_{$flname}_{$pname}";
            try {
                $frame = Frame::buildFrame($ftype, $flags, $payload);
                $phpWire = $frame->toBytes();
                $goldenWire = loadGoldenFrame($name);
                if ($goldenWire === null) {
                    test($name, 'frame_binary_match', false, "golden not found");
                    test($name, 'frame_parse_golden', false, "golden not found");
                    continue;
                }
                $ok = $phpWire === $goldenWire;
                test($name, 'frame_binary_match', $ok,
                    $ok ? null : sprintf("PHP(%dB):%s != PY(%dB):%s",
                        strlen($phpWire), bin2hex($phpWire), strlen($goldenWire), bin2hex($goldenWire)));
                $result = Frame::parseFrame($goldenWire, 0);
                if ($result instanceof ParseResultComplete) {
                    $pok = $result->frame->frameType === $ftype
                        && $result->frame->flags === $flags
                        && $result->frame->payload === $payload;
                    test($name, 'frame_parse_golden', $pok);
                } else {
                    test($name, 'frame_parse_golden', false, 'not complete: ' . $result::class);
                }
            } catch (\Throwable $e) {
                test($name, 'frame_binary_match', false, $e::class . ': ' . $e->getMessage());
                test($name, 'frame_parse_golden', false, $e::class . ': ' . $e->getMessage());
            }
        }
    }
}

// ═══ 7. Compressed Frame Integration ═══
echo "\n7. Compressed Frame Integration\n";
$integrationPayloads = [
    ['initialize', ['jsonrpc' => '2.0', 'id' => 1, 'method' => 'initialize',
                    'params' => ['protocolVersion' => '2025-06-18']]],
    ['tools_list', ['jsonrpc' => '2.0', 'id' => 2, 'result' => [
        'tools' => [['name' => 'search', 'description' => 'Search code',
                      'inputSchema' => ['type' => 'object', 'properties' => ['query' => ['type' => 'string']]]]]
    ]]],
];

foreach ($integrationPayloads as [$pname, $payload]) {
    $name = "integration_$pname";
    try {
        $compressed = Compress::compress($payload);
        $frame = Frame::buildFrame(Frame::REQUEST, Frame::FLAG_COMPRESSED, $compressed);
        $wire = $frame->toBytes();
        $result = Frame::parseFrame($wire, 0);
        if ($result instanceof ParseResultComplete) {
            $frameOk = $result->frame->isCompressed();
            $decompressed = Compress::decompress($result->frame->payload);
            $payloadOk = jsonEqual($decompressed, $payload);
            test($name, 'integration_roundtrip', $frameOk && $payloadOk);

            $goldenWire = loadGoldenFrame($name);
            if ($goldenWire !== null) {
                $match = $wire === $goldenWire;
                test($name, 'integration_binary_match', $match,
                    $match ? null : sprintf("PHP(%dB):%s != PY(%dB):%s",
                        strlen($wire), bin2hex($wire), strlen($goldenWire), bin2hex($goldenWire)));

                $pyResult = Frame::parseFrame($goldenWire, 0);
                if ($pyResult instanceof ParseResultComplete && $pyResult->frame->isCompressed()) {
                    $pyDecompressed = Compress::decompress($pyResult->frame->payload);
                    test($name, 'integration_parse_python', jsonEqual($pyDecompressed, $payload));
                } else {
                    test($name, 'integration_parse_python', false, 'failed to parse PY integration frame');
                }
            } else {
                test($name, 'integration_binary_match', false, "golden not found");
                test($name, 'integration_parse_python', false, "golden not found");
            }
        } else {
            test($name, 'integration_roundtrip', false, 'not complete: ' . $result::class);
        }
    } catch (\Throwable $e) {
        test($name, 'integration_roundtrip', false, $e::class . ': ' . $e->getMessage());
        test($name, 'integration_binary_match', false, $e::class . ': ' . $e->getMessage());
        test($name, 'integration_parse_python', false, $e::class . ': ' . $e->getMessage());
    }
}

// ═══ Summary ═══
echo "\n══════════════════════════════════════════════════\n";
printf("Results: %d passed, %d failed, %d total\n", $passed, $failed, $passed + $failed);
exit($failed === 0 ? 0 : 1);

File: implementations/rust/tests/e2e_test.rs
//! LUMEN e2e cross-implementation test — Rust.
//!
//! Validates that the Rust implementation produces binary output
//! identical to the Python golden binaries, and that cross-decoding works.
//!
//! Run: cargo test --test e2e_test

use lumen::compress::{compress, decompress};
use lumen::hyb128::{encode as hyb128_encode, decode as hyb128_decode};
use lumen::frame::{self, build, build_size, parse, ParseResult};

use serde_json::Value;
use std::fs;
use std::path::PathBuf;

// ── Helpers ────────────────────────────────────────────────────────────────

fn vectors_path() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..").join("..")
        .join("tests").join("e2e").join("shared_vectors.json")
}

fn golden_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..").join("..")
        .join("tests").join("e2e").join("golden")
}

fn load_vectors() -> Vec<(String, Value)> {
    let raw = fs::read_to_string(vectors_path()).expect("shared_vectors.json not found");
    let data: Value = serde_json::from_str(&raw).expect("Invalid JSON");
    data["vectors"].as_array().unwrap().iter().map(|v| {
        (v["name"].as_str().unwrap().to_string(), v["value"].clone())
    }).collect()
}

fn load_golden(name: &str) -> Option<Vec<u8>> {
    fs::read(golden_dir().join(format!("{}.lumen", name))).ok()
}

fn load_golden_frame(name: &str) -> Option<Vec<u8>> {
    fs::read(golden_dir().join(format!("{}.frame", name))).ok()
}

fn skip(name: &str) -> bool {
    matches!(name, "int_large")
}

// ═══════════════════════════════════════════════════════════════════════════════
// 1. Compress roundtrip
// ═══════════════════════════════════════════════════════════════════════════════

#[test]
fn compress_roundtrip_all() {
    for (name, value) in load_vectors() {
        if skip(&name) { continue; }
        let compressed = compress(&value);
        let decompressed = decompress(&compressed)
            .unwrap_or_else(|| panic!("decompress returned None for {}", name));
        assert_eq!(
            serde_json::to_string(&value).unwrap(),
            serde_json::to_string(&decompressed).unwrap(),
            "roundtrip mismatch for {}", name
        );
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// 2. Semantic compatibility with Python golden files
//    Note: raw binary may differ due to object key ordering (Rust sorts,
//    Python preserves insertion order). Both are valid LUMEN.
// ═══════════════════════════════════════════════════════════════════════════════

#[test]
fn semantic_compat_python_golden() {
    for (name, value) in load_vectors() {
        if skip(&name) { continue; }
        let golden = match load_golden(&name) {
            Some(g) => g,
            None => continue,
        };

        // Decompress Python golden and compare with our decompress of our compress
        let py_decoded = decompress(&golden)
            .unwrap_or_else(|| panic!("Failed to decompress Python golden for {}", name));
        let our_compressed = compress(&value);
        let our_decoded = decompress(&our_compressed)
            .unwrap_or_else(|| panic!("Failed to decompress own output for {}", name));

        assert_eq!(
            serde_json::to_string(&py_decoded).unwrap(),
            serde_json::to_string(&our_decoded).unwrap(),
            "Semantic mismatch \"{}\": Python and Rust decodes differ", name
        );
        assert_eq!(
            serde_json::to_string(&value).unwrap(),
            serde_json::to_string(&py_decoded).unwrap(),
            "Semantic mismatch \"{}\": Python golden doesn't roundtrip to original", name
        );
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// 3. Rust decodes Python golden binaries
// ═══════════════════════════════════════════════════════════════════════════════

#[test]
fn decode_python_golden() {
    for (name, value) in load_vectors() {
        if skip(&name) { continue; }
        let golden = match load_golden(&name) {
            Some(g) => g,
            None => continue,
        };
        let decompressed = decompress(&golden)
            .unwrap_or_else(|| panic!("Rust failed to decode Python golden for {}", name));
        assert_eq!(
            serde_json::to_string(&value).unwrap(),
            serde_json::to_string(&decompressed).unwrap(),
            "decode Python golden mismatch for {}", name
        );
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// 4. Binary stability
// ═══════════════════════════════════════════════════════════════════════════════

#[test]
fn binary_stability() {
    let cases: Vec<(&str, Value)> = vec![
        ("null", Value::Null),
        ("bool", Value::Bool(true)),
        ("int", serde_json::json!(42)),
        ("float", serde_json::json!(3.14)),
        ("string", Value::String("hello".into())),
        ("array", serde_json::json!([1, 2, 3])),
        ("object", serde_json::json!({"key": "value"})),
        ("mcp_init", serde_json::json!({"jsonrpc": "2.0", "method": "initialize"})),
    ];
    for (name, value) in &cases {
        let c1 = compress(value);
        let c2 = compress(value);
        assert_eq!(c1, c2, "Non-deterministic compression for {}", name);
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// 5. Hyb128 roundtrip
// ═══════════════════════════════════════════════════════════════════════════════

#[test]
fn hyb128_roundtrip() {
    let cases = [(0u64, 1usize), (1, 1), (42, 1), (63, 1),
                 (64, 3), (255, 3), (1000, 3), (65535, 3),
                 (65536, 5), (100000, 5), (1000000, 5)];
    for (value, expected) in cases {
        let mut buf = [0u8; 11];
        let n = hyb128_encode(value, &mut buf);
        assert_eq!(n, expected, "encode({}) = {}, expected {}", value, n, expected);
        let dec = hyb128_decode(&buf[..n]).expect("decode returned None");
        assert_eq!(dec.value, value);
        assert_eq!(dec.header_len, expected);
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// 6. Frame roundtrip
// ═══════════════════════════════════════════════════════════════════════════════

#[test]
fn frame_roundtrip() {
    let json_small = serde_json::to_vec(&serde_json::json!({"method":"ping"})).unwrap();
    let json_mcp = serde_json::to_vec(&serde_json::json!({
        "jsonrpc":"2.0","id":1,"method":"initialize",
        "params":{"protocolVersion":"2025-06-18"}
    })).unwrap();

    let payloads: Vec<(&str, &[u8])> = vec![
        ("empty", &[]),
        ("hello", b"hello"),
        ("json_small", &json_small),
        ("json_mcp", &json_mcp),
    ];
    let ftypes: [(u8, &str); 3] = [
        (frame::TYPE_REQUEST, "REQUEST"),
        (frame::TYPE_RESPONSE, "RESPONSE"),
        (frame::TYPE_NOTIFY, "NOTIFY"),
    ];
    let flagsets: [(u8, &str); 3] = [
        (0, "none"),
        (frame::FLAG_COMPRESSED, "compressed"),
        (frame::FLAG_PRIORITY, "priority"),
    ];

    for (pname, payload) in &payloads {
        for (ftype, tname) in &ftypes {
            for (flags, fname) in &flagsets {
                let name = format!("frame_{}_{}_{}", tname, fname, pname);
                let mut buf = vec![0u8; build_size(payload.len())];
                let n = build(*ftype, *flags, payload, &mut buf);
                assert_eq!(n, buf.len());
                match parse(&buf) {
                    ParseResult::Complete { frame, consumed } => {
                        assert_eq!(frame.frame_type, *ftype, "type: {}", name);
                        assert_eq!(frame.flags, *flags, "flags: {}", name);
                        assert_eq!(frame.payload, *payload, "payload: {}", name);
                        assert_eq!(consumed, n, "consumed: {}", name);
                    }
                    other => panic!("Expected Complete for {}, got {:?}", name, other),
                }
            }
        }
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// 7. Frame semantic match with Python golden frames
//    Raw bytes may differ due to JSON key ordering in payloads.
//    We parse both and compare frame metadata + payload content.
// ═══════════════════════════════════════════════════════════════════════════════

#[test]
fn frame_semantic_match_python() {
    let json_small = serde_json::to_vec(&serde_json::json!({"method":"ping"})).unwrap();
    let json_mcp = serde_json::to_vec(&serde_json::json!({
        "jsonrpc":"2.0","id":1,"method":"initialize",
        "params":{"protocolVersion":"2025-06-18"}
    })).unwrap();

    let payloads: Vec<(&str, &[u8])> = vec![
        ("empty", &[]),
        ("hello", b"hello"),
        ("json_small", &json_small),
        ("json_mcp", &json_mcp),
    ];
    let ftypes: [(u8, &str); 3] = [
        (frame::TYPE_REQUEST, "REQUEST"),
        (frame::TYPE_RESPONSE, "RESPONSE"),
        (frame::TYPE_NOTIFY, "NOTIFY"),
    ];
    let flagsets: [(u8, &str); 3] = [
        (0, "none"),
        (frame::FLAG_COMPRESSED, "compressed"),
        (frame::FLAG_PRIORITY, "priority"),
    ];

    for (pname, payload) in &payloads {
        for (ftype, tname) in &ftypes {
            for (flags, fname) in &flagsets {
                let name = format!("frame_{}_{}_{}", tname, fname, pname);

                let golden = match load_golden_frame(&name) {
                    Some(g) => g,
                    None => continue,
                };

                // Parse Python golden frame
                match parse(&golden) {
                    ParseResult::Complete { frame: py_frame, .. } => {
                        assert_eq!(py_frame.frame_type, *ftype,
                            "type mismatch parsing Python golden {}", name);
                        assert_eq!(py_frame.flags, *flags,
                            "flags mismatch parsing Python golden {}", name);

                        // For compressed JSON payloads, decompress and compare semantically
                        if py_frame.flags & frame::FLAG_COMPRESSED != 0 && pname.starts_with("json") {
                            // Check if payload is actually compressed LUMEN (starts with tag 0xE0-0xE7)
                            let is_compressed = !py_frame.payload.is_empty()
                                && (py_frame.payload[0] & 0xF8) == 0xE0;
                            if is_compressed {
                                let py_val = decompress(py_frame.payload)
                                    .unwrap_or_else(|| panic!("{}: Python compressed payload failed to decompress", name));
                                let our_val: Value = serde_json::from_slice(payload)
                                    .unwrap_or_else(|e| panic!("{}: Our payload not JSON: {}", name, e));
                                assert_eq!(
                                    serde_json::to_string(&py_val).unwrap(),
                                    serde_json::to_string(&our_val).unwrap(),
                                    "compressed payload semantic mismatch for {}", name
                                );
                            } else {
                                // Flag set but payload is raw JSON — compare semantically
                                let py_val: Value = serde_json::from_slice(py_frame.payload)
                                    .unwrap_or_else(|e| panic!("{}: Python payload not JSON: {}", name, e));
                                let our_val: Value = serde_json::from_slice(payload)
                                    .unwrap_or_else(|e| panic!("{}: Our payload not JSON: {}", name, e));
                                assert_eq!(
                                    serde_json::to_string(&py_val).unwrap(),
                                    serde_json::to_string(&our_val).unwrap(),
                                    "payload semantic mismatch for {}", name
                                );
                            }
                        } else if pname.starts_with("json") {
                            // Uncompressed JSON payloads — normalize and compare
                            let py_val: Value = serde_json::from_slice(py_frame.payload)
                                .unwrap_or_else(|e| panic!("{}: Python payload not JSON: {}", name, e));
                            let our_val: Value = serde_json::from_slice(payload)
                                .unwrap_or_else(|e| panic!("{}: Our payload not JSON: {}", name, e));
                            assert_eq!(
                                serde_json::to_string(&py_val).unwrap(),
                                serde_json::to_string(&our_val).unwrap(),
                                "payload semantic mismatch for {}", name
                            );
                        } else {
                            // Non-JSON payloads should match exactly
                            assert_eq!(py_frame.payload, *payload,
                                "payload mismatch for {}", name);
                        }
                    }
                    other => panic!("parse Python golden {}: {:?}", name, other),
                }
            }
        }
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// 8. Compressed frame integration (semantic comparison)
// ═══════════════════════════════════════════════════════════════════════════════

#[test]
fn compressed_frame_integration() {
    let payloads: Vec<(&str, Value)> = vec![
        ("initialize", serde_json::json!({
            "jsonrpc":"2.0","id":1,"method":"initialize",
            "params":{"protocolVersion":"2025-06-18"}
        })),
        ("tools_list", serde_json::json!({
            "jsonrpc":"2.0","id":2,"result":{"tools":[{
                "name":"search","description":"Search code",
                "inputSchema":{"type":"object","properties":{"query":{"type":"string"}}}
            }]}
        })),
    ];

    for (pname, payload) in &payloads {
        let name = format!("integration_{}", pname);

        let compressed = compress(payload);
        let mut buf = vec![0u8; build_size(compressed.len())];
        build(frame::TYPE_REQUEST, frame::FLAG_COMPRESSED, &compressed, &mut buf);

        // Parse + decompress own frame
        match parse(&buf) {
            ParseResult::Complete { frame, .. } => {
                assert_eq!(frame.flags, frame::FLAG_COMPRESSED);
                let dec = decompress(frame.payload)
                    .unwrap_or_else(|| panic!("decompress failed for {}", name));
                assert_eq!(
                    serde_json::to_string(payload).unwrap(),
                    serde_json::to_string(&dec).unwrap(),
                    "integration decompress mismatch for {}", name
                );
            }
            other => panic!("parse {}: {:?}", name, other),
        }

        // Parse + decompress Python golden (semantic compare)
        if let Some(py_golden) = load_golden_frame(&name) {
            match parse(&py_golden) {
                ParseResult::Complete { frame, .. } => {
                    assert_eq!(frame.flags, frame::FLAG_COMPRESSED,
                        "Python golden flags mismatch for {}", name);
                    let dec = decompress(frame.payload)
                        .unwrap_or_else(|| panic!("decompress Python golden failed for {}", name));
                    assert_eq!(
                        serde_json::to_string(payload).unwrap(),
                        serde_json::to_string(&dec).unwrap(),
                        "integration: decompress Python golden mismatch for {}", name
                    );
                }
                other => panic!("parse Python golden {}: {:?}", name, other),
            }
        }
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// 9. Streaming frame parse (Rust equivalent of FrameAssembler)
// ═══════════════════════════════════════════════════════════════════════════════

#[test]
fn streaming_frame_parse() {
    // Two frames in one buffer
    let p1 = b"A";
    let p2 = b"BB";
    let mut b1 = vec![0u8; build_size(p1.len())];
    let mut b2 = vec![0u8; build_size(p2.len())];
    build(frame::TYPE_NOTIFY, 0, p1, &mut b1);
    build(frame::TYPE_RESPONSE, 0, p2, &mut b2);
    let combined: Vec<u8> = b1.iter().chain(b2.iter()).copied().collect();

    match parse(&combined) {
        ParseResult::Complete { frame, consumed } => {
            assert_eq!(frame.payload, p1);
            let rest = &combined[consumed..];
            match parse(rest) {
                ParseResult::Complete { frame, .. } => {
                    assert_eq!(frame.payload, p2);
                }
                other => panic!("Second frame: {:?}", other),
            }
        }
        other => panic!("First frame: {:?}", other),
    }

    // Truncated frame → IncompletePayload (header parsed, payload missing)
    let payload = b"chunked_test";
    let mut buf = vec![0u8; build_size(payload.len())];
    build(frame::TYPE_REQUEST, frame::FLAG_COMPRESSED, payload, &mut buf);
    let mid = buf.len() / 2;
    assert!(matches!(parse(&buf[..mid]), ParseResult::IncompletePayload { .. }));
    assert!(matches!(parse(&buf), ParseResult::Complete { .. }));
}
